Add cartridges overview in explorer

taurgis opened the pull request:
![screenshot_cartridges](https://user-images.githubusercontent.com/2530193/28382770-8f8aeb80-6cbf-11e7-8e8c-5634d5b536d9.png)

Started development on a cartridge overview. Still in development, so wouldn't recommend accepting the pull request just yet.

Just wanted to let you know what i'm up to.

taurgis commented:
@SqrTT This is now mergeable (if it is ok for you). I have removed some unused code and refactored it. 

Also I have added some file icons and right clicking on a cartridge in the explorer will show a context menu to force a  clean/uploadall.

This is just a stepping stone to whats possible here off course: Later one it could be possible to add context menu's to add specific file types and put them in to the correct folder or even create a cartridge.

SqrTT commented on src/providers/CartridgesView.ts:
```diff
@@ -0,0 +1,146 @@
+
+	private getCartridgesInWorkspace(workspaceRoot: string): CartridgeItem[] {
+		if (this.pathExists(workspaceRoot)) {
+			var directories = this.getDirectories(workspaceRoot)
```
Not always cartridges root and `workspaceRoot` is the same folder. For example in my company we use next structure:

```
├── cartridges <-- contains all cartridges
├── dw.json
├── dw.json.example
├── gulpfile.js
├── node_modules
├── package.json
├── package-lock.json
├── README.md
├── ssl
└── types
```
Hence need the better way to detect that case.

PS. Also, would be great considered `modules` folder as cartridge too.

taurgis replied:
A possible solution to find the cartridges, no matter where they are is to look for all .project files with 

`	<natures>
		<nature>com.demandware.studio.core.beehiveNature</nature>
	</natures>
`

SqrTT commented:
Great to see someone input. I'll take a look.

taurgis commented:
Used package "find" to look for .project files with a regex check on the contents to see if its a Salesforce (demandware) cartridge.

taurgis commented:
Added:
 - Look in entire workspace for cartridges (no matter where they are)
 - Support modules
 - Added command to refresh the treeview
 - Added command to create a cartridge

SqrTT commented on CHANGELOG.md:
```diff
@@ -12,7 +12,7 @@
 * Added quick open for custom tags
 * Added `extension.prophet.cartridges.path` property that allows quick open don't ask a user to choose the file.
-* Added `extension.prophet.ismlServer.activateOn` property that allow activate isml server for non standatd (isml) files, ex. `html`
+* Added `extension.prophet.ismlServer.activateOn` property that allow activate isml server for non standar1d (isml) files, ex. `html`
```
`standar1d` -> `standatd`

SqrTT commented on package.json:
```diff
@@ -28,6 +28,7 @@
     },
     "dependencies": {
         "chokidar": "^1.7.0",
+        "find": "^0.2.7",
```
Is not `glob` solve the same issue?

taurgis replied:
Yep it does, made the required changes. Didn't notice that glob dit something similar (haven't used the package before).

SqrTT commented on src/providers/CartridgesView.ts:
```diff
@@ -0,0 +1,147 @@
+				let fileContent = fs.readFileSync(projectFile, 'UTF-8');
+
+				// Use a regex check rather than parsing the XML since the file is not that big.
+				var nature = fileContent.match(/<nature>com.demandware.studio.core.beehiveNature<\/nature>/);
```
Probably `fileContent.includes('com.demandware.studio.core.beehiveNature')` would be faster

SqrTT commented on src/lib/FileHelper.ts:
```diff
@@ -0,0 +1,23 @@
+
+export function getFiles(srcpath) {
+    return fs.readdirSync(srcpath)
+        .filter(file => !fs.lstatSync(path.join(srcpath, file)).isDirectory())
```
`isFile`?

taurgis replied:
Changed!

SqrTT commented on src/lib/FileHelper.ts:
```diff
@@ -0,0 +1,23 @@
+'use strict';
```
Is there a way to do this job asynchronously?

taurgis replied:
Which job do you mean?

taurgis replied:
I have made the getfiles and getdirectories an async function using promises.

SqrTT commented on src/providers/CartridgesView.ts:
```diff
@@ -0,0 +1,147 @@
+
+	private getCartridgesInWorkspace(workspaceRoot: string): CartridgeItem[] {
+		if (pathExists(workspaceRoot)) {
+			var projectFiles = find.fileSync(/\.project/, workspaceRoot);
```
Basically, we can find out cartridges root based on `dw.json` and it's config.

taurgis replied:
Yes, but only if it exists. Currently its only required if you want to make use of the upload feature.
